test/grouping.jl

```diff
@@ -2523,4 +2538,114 @@ end
+    @test gdf == groupby_checked(df, [:a, :b])
+
+    df = DataFrame(c=1, b=2, d=3, a=4)
+    gdf = groupby_checked(df, [:a, :b])
+    @test transform!(gdf, :c => :e, ungroup=false) === gdf
+    @test df == DataFrame(c=1, b=2, d=3, a=4, e=1)
+    @test gdf == groupby_checked(df, [:a, :b])
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    dc_gdf = deepcopy(gdf)
+    df_res = select(gdf, :c => :e)
+    @test df_res == DataFrame(a=Char[], b=Float64[], e=String[])
+    @test eltype.(eachcol(df_res)) == [Char, Float64, String]
+    @test gdf == dc_gdf
+    @test select(gdf, :c => :e, ungroup=false) ==
+          groupby_checked(DataFrame(a=Char[], b=Float64[], e=String[]), [:a, :b])
+    @test select(gdf, keepkeys=false) == DataFrame()
+    df_res = select(gdf, :c => :e, keepkeys=false)
+    @test df_res == DataFrame(e=String[])
+    @test eltype(df_res.e) == String
+    @test gdf == dc_gdf
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    @test select!(gdf, :c => :e) === df
+    @test df == DataFrame(a=Char[], b=Float64[], e=String[])
+    @test gdf == groupby_checked(df, [:a, :b])
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    @test select!(gdf, :c => :e, ungroup=false) === gdf
+    @test df == DataFrame(a=Char[], b=Float64[], e=String[])
+    @test gdf == groupby_checked(df, [:a, :b])
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    dc_gdf = deepcopy(gdf)
+    @test transform(gdf, :c => :e) ==
+          DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[], e=String[])
+    @test gdf == dc_gdf
+    @test transform(gdf, :c => :e, ungroup=false) ==
+          groupby_checked(DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[], e=String[]), [:a, :b])
+    @test transform(gdf, :c => :e, keepkeys=false) ==
+          DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[], e=String[])
+    @test gdf == dc_gdf
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    @test transform!(gdf, :c => :e) === df
+    @test df == DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[], e=String[])
+    @test gdf == groupby_checked(df, [:a, :b])
+
+    df = DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[])
+    gdf = groupby_checked(df, [:a, :b])
+    @test transform!(gdf, :c => :e, ungroup=false) === gdf
+    @test df == DataFrame(c=String[], b=Float64[], d=Bool[], a=Char[], e=String[])
+    @test gdf == groupby_checked(df, [:a, :b])
+end
+
+@testset "combine on empty data frame" begin
+    df = DataFrame(x=Int[])
+    @test combine(df, nrow) == DataFrame(nrow=0)
+    @test combine(df, nrow => :z) == DataFrame(z=0)
+    @test combine(df, [nrow => :z]) == DataFrame(z=0)
+    @test combine(df, :x => (x -> 1:2) => :y) == DataFrame(y=1:2)
```

Comment:
Maybe test a function which relies on the input column type, and check that the resulting type is correct?

Reply:
added